src/hotspot/share/ci/ciEnv.cpp

```diff
@@ -711,35 +749,29 @@ ciConstant ciEnv::get_constant_by_index_impl(const constantPoolHandle& cpool,
     return ciConstant((jdouble)cpool->double_at(index));
   } else if (tag.is_string()) {
     EXCEPTION_CONTEXT;
-    oop string = NULL;
-    assert(cache_index >= 0, "should have a cache index");
-    string = cpool->string_at(index, cache_index, THREAD);
+    assert(obj_index >= 0, "should have an object index");
+    oop string = cpool->string_at(index, obj_index, THREAD);
     if (HAS_PENDING_EXCEPTION) {
       CLEAR_PENDING_EXCEPTION;
       record_out_of_memory_failure();
       return ciConstant();
     }
-    ciObject* constant = get_object(string);
-    if (constant->is_array()) {
-      return ciConstant(T_ARRAY, constant);
-    } else {
-      assert (constant->is_instance(), "must be an instance, or not? ");
-      return ciConstant(T_OBJECT, constant);
-    }
+    ciInstance* constant = get_object(string)->as_instance();
+    return ciConstant(T_OBJECT, constant);
```

Comment:
Please explain what happens for array constants now.

Reply:
It is part of `JVM_CONSTANT_String` resolution, so string constant is expected here.